### PR TITLE
Utility functions for iterative algorithms

### PR DESCRIPTION
**Pull request by uellue**
* Fast image transforms using a sparse matrix product
* Coordinate transforms to match detector data with forward model
* Calculate stacks of shifted and rolled products of a smaller illumination within a larger object, including sub-pixel shifting
* Aggregate shifted and rolled updates within a larger object

The functions can perform FFT shifts on the fly to reduce memory transfers. Both CPU and CUDA are supported. They are tested to conform precisely to the conventions of `libertem.api.create_com_analysis()` and also `SSB_UDF`.

CC @bangunarya @Sniper2k that's the stuff I have mentioned recently. If you'd like to try it, let me know! :-)

## Contributor Checklist:

* [x] I have added myself to [the creators.json file](https://github.com/Ptychography-4-0/ptychography/blob/master/packaging/creators.json)
* [x] I have added [a changelog entry](https://github.com/Ptychography-4-0/ptychography/tree/master/docs/source/changelog) for my contribution
* [x] I have added/updated documentation for all user-facing changes
* [x] I have added/updated test cases

### CODE REVIEW

File: src/ptychography40/reconstruction/common.py [hunk 1]
```diff
@@ -126,3 +132,543 @@ def bounding_box(array):
...
+    Parameters
+    ----------
+    probe : numpy.ndarray
+    bins : int or tuple(int)
+        Number of antialiasing steps in y and x axis. Can be int as well
+
+    Returns
+    -------
+    probes : numpy.ndarray
+        4D, dhape bins + probe.shape or (bins, bins) + probe.shape
```
Comment (uellue) [inline src/ptychography40/reconstruction/common.py:487]:
```suggestion
        4D, shape bins + probe.shape or (bins, bins) + probe.shape
```

**Comment by SimeonEhrig** [comment]
@uellue Is there are a rendered version of the sphinx doc available? In the alpaka-group, we use readthedoc. There is a CI job, which build the sphinx doc and provides an link to the result.

**Comment by uellue** [comment]
> @uellue Is there are a rendered version of the sphinx doc available? In the alpaka-group, we use readthedoc. There is a CI job, which build the sphinx doc and provides an link to the result.

At the moment the [rendered documentation](https://ptychography-4-0.github.io/ptychography/) is only available after merging or by running `tox -e docs` locally. @sk1p, having the rendered docs of a PR available could help with reviewing, since sometimes it builds cleanly but looks terrible. I've reviewed the rendered docs locally up to now, which is a bit tedious.

**Comment by SimeonEhrig** [comment]
> At the moment the [rendered documentation](https://ptychography-4-0.github.io/ptychography/) is only available after merging or by running `tox -e docs` locally. @sk1p, having the rendered docs of a PR available could help with reviewing, since sometimes it builds cleanly but looks terrible. I've reviewed the rendered docs locally up to now, which is a bit tedious.

Thanks. I was not sure, if the CI job provides an artifact of the build. But building locally is also fine.

**Review by SimeonEhrig — APPROVED** [review]

**Comment by uellue** [comment]
@sk1p is the notebook that shows the transforms understandable? Do we need more documentation, are there other things to address? It would be nice to merge this some time next week to make it easier to use in downstream code. :-)

**Review by sk1p — APPROVED** [review]

**Comment by sk1p** [comment]
> is the notebook that shows the transforms understandable?

Yes, thanks! 

> Do we need more documentation, are there other things to address? It would be nice to merge this some time next week to make it easier to use in downstream code. :-)

I think documentation can always be improved, but let's merge this now (perfect is the enemy of good, and all that...)

**Pull request merged**